File: editor_frontend/src/models/Exhibit.ts
import firebase from "firebase/compat";
import DocumentData = firebase.firestore.DocumentData;
import exhibitEditor from "@/components/ExhibitEditor.vue";

export class Exhibit {
  id: string; // id of exhibit
  name: string; // name of exhibit
  description: string; //short description
  extra: ExhibitPair[]; //extra parameters
  widgets: Widget[];
  icon: string; // path to image with icon
  location: string; // path to image with location on exhibition

  constructor();
  constructor(id: string, name: string);
  constructor(id?: string, name?: string) {
    this.id = id || "";
    this.name = name || "";
    this.description = "";
    this.extra = [];
    this.widgets = [];
    this.icon = "";
    this.location = "";
  }

  toFirestoreConverter() {
    // const pairArr: ExhibitPair[] = [];
    // Array.from(this.extra.entries()).forEach((el) => {
    //   pairArr.push(new ExhibitPair(el[0], el[1]));
    // });

    return {
      name: this.name,
      description: this.description,
      extra: JSON.stringify(this.extra),
      widgets: JSON.stringify(this.widgets),
    };
  }

  static fromFirebaseConverter(id: string, data: DocumentData | undefined) {
    const newExhibit = new Exhibit();
    newExhibit.id = id;
    newExhibit.name = data?.name;
    newExhibit.description = data?.description;
    newExhibit.widgets = JSON.parse(data?.widgets);
    newExhibit.extra = JSON.parse(data?.extra);

    // const pairArr: ExhibitPair[] = JSON.parse(data?.extra);
    // pairArr.forEach((el) => {
    //   newExhibit.extra.set(el.value, el.key);
    // });

    return newExhibit;
  }
}

export class Widget {
  id: number;
  type: string;
  title: string;
  data: string;
  imagesURLs: string[];

  constructor(id: number, type: string) {
    this.id = 0;
    this.type = "";
    this.title = "";
    this.data = "";
    this.imagesURLs = [];
  }

  addImageURL(name: string){
    this.imagesURLs.push(name);
  }
}

export class ExhibitPair {
  value: string;
  key: string;
  linkId: number;

  constructor(value: string, key: string);
  constructor(value: string, key: string, linkId: number);
  constructor(value: string, key: string, linkId?: number) {
    this.value = value;
    this.key = key;
    this.linkId = linkId || 0;
  }
}

File: androidApp/app/src/main/java/pl/edu/pg/cloudlib/database/Exhibit.kt
package pl.edu.pg.cloudlib.database

import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.QueryDocumentSnapshot
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import org.json.JSONArray

class Exhibit {
    lateinit var id: String; // id of exhibit
    lateinit var name: String; // name of exhibit
    lateinit var description: String; //short description
    lateinit var extra: Array<ExhibitPair>; //extra parameters
    lateinit var widgets: Array<Widget>;
    lateinit var icon: String; // path to image with icon
    lateinit var location: String; // path to image with location on exhibition

    fun init(
        /*id: String,
        name: String,
        description: String,
        extra: Array<ExhibitPair>,
        widgets: Array<Widget>,
        icon: String,
        location: String*/
    ) {
        this.id = "id"
        this.name = "name"
        this.description = "description"
        this.extra = arrayOf()//extra
        this.widgets = arrayOf()//widgets
        this.icon = "icon"
        this.location = "location"
    }

    override fun toString(): String {
        return "Exhibit(id='$id', name='$name', description='$description', extra=${extra.contentToString()}, widgets=${widgets.contentToString()}, icon='$icon', location='$location')"
    }

    companion object {
        fun fromFirebaseConverter(doc: DocumentSnapshot): Exhibit {
            val newExhibit = Exhibit()
            newExhibit.id = doc.id
            newExhibit.name = doc.data?.get("name").toString()
            newExhibit.description = doc.data?.get("description").toString()
            newExhibit.widgets = Gson().fromJson(doc.data?.get("widgets").toString(), object : TypeToken<Array<Widget>>() {}.type)?: arrayOf()
            newExhibit.extra = Gson().fromJson(doc.data?.get("extra").toString(), object : TypeToken<Array<ExhibitPair>>() {}.type)?: arrayOf()
            newExhibit.icon = doc.data?.get("icon").toString()
            newExhibit.location = doc.data?.get("location").toString()

            return newExhibit
        }
    }

}

class Widget {
    var id: Int = 0;
    lateinit var type: String;
    lateinit var title: String;
    lateinit var data: String;
    lateinit var imagesURLs: Array<String>;

    fun init(
        id: Int,
        type: String,
        title: String,
        data: String,
        imagesURLs: Array<String>
    ) {
        this.id = id;
        this.type = type;
        this.title = title;
        this.data = data;
        this.imagesURLs = imagesURLs;
    }
}

class ExhibitPair {
    lateinit var value: String;
    lateinit var key: String;
    lateinit var linkId: String;

    fun init(value: String, key: String, linkId: String) {
        this.value = value;
        this.key = key;
        this.linkId = linkId;
    }
}